client/material/tooltip.tsx

```diff
@@ -0,0 +1,121 @@
+import React, { useState } from 'react'
+import { useTransition, UseTransitionProps } from 'react-spring'
+import styled from 'styled-components'
+import { background100 } from '../styles/colors'
+import { caption } from '../styles/typography'
+import { OriginX, OriginY, PopoverContent, useAnchorPosition } from './popover'
+import { Portal } from './portal'
+import { defaultSpring } from './springs'
+
+const isDev = __WEBPACK_ENV.NODE_ENV !== 'production'
+
+const Container = styled.div`
+  ${caption};
+  display: flex;
+  justify-content: center;
+  align-items: center;
+
+  min-height: 24px;
+  padding: 4px 8px;
+  border-radius: 4px;
+  background-color: ${background100};
+`
+
+export type TooltipPosition = 'left' | 'right' | 'top' | 'bottom'
+
+interface TooltipProps {
+  /** The text that should be displayed in the Tooltip. */
+  text: string
+  /** The child element that the Tooltip should be linked to. Should only be a single element. */
+  children: React.ReactNode
+  /** One of the four sides that can be used to position the Tooltip. Defaults to 'bottom'. */
+  position?: TooltipPosition
+  /** Class name applied to the root container of the Tooltip (that has the background, etc.). */
+  className?: string
+}
+
+/**
+ * A component that displays some content in a floated UI element when a user hovers over a target
+ * element. Should generally be used over native `title` attribute on elements.
+ *
+ * Utilizes popovers to deal with positioning, but with a much simpler API than that of a Popover.
+ */
+export function Tooltip({ text, children, position = 'bottom', className }: TooltipProps) {
+  const [anchorElem, setAnchorElem] = useState<HTMLElement>()
+  const transition = useTransition<boolean, UseTransitionProps<boolean>>(!!anchorElem, {
+    from: { opacity: 0, scale: 0.667 },
+    enter: { opacity: 1, scale: 1 },
+    leave: { opacity: 0, scale: 0.333 },
+    delay: 800,
+    config: (item, index, phase) => key =>
+      phase === 'leave' || key === 'opacity' ? { ...defaultSpring, clamp: true } : defaultSpring,
+  })
+
+  const anchorOriginX = position === 'top' || position === 'bottom' ? 'center' : position
+  const anchorOriginY = position === 'left' || position === 'right' ? 'center' : position
+  let [, anchorX = 0, anchorY = 0] = useAnchorPosition(
+    anchorOriginX,
+    anchorOriginY,
+    anchorElem ?? null,
+  )
+
+  const onMouseEnter = (event: React.MouseEvent) => {
+    setAnchorElem(event.currentTarget as HTMLElement)
+  }
+  const onMouseLeave = (event: React.MouseEvent) => {
+    setAnchorElem(undefined)
+  }
+
+  const childrenCount = React.Children.count(children)
```

Comment:
So *if* you wanted to do this, React already has a thing for this, don't really need to build your own: https://reactjs.org/docs/react-api.html#reactchildrenonly

However I think this is a bad idea. I don't see why you'd ever need to do this, and it's not type-safe for you to do what you're doing with the props here (you have no restrictions on the passed in children to ensure they can take these props).

A better way imo is to wrap the children and attach these event handlers to that element. I would apply this components `className` to that wrapper and allow a custom `PopoverContent` replacement to be passed in if they need to customize the tooltip style (but honestly that should be extremely rare, these things are probably very consistent)

Reply:
Another alternative to this approach, if you want to avoid the wrapper, is to split the component into two parts: the UI and a hook. The UI is mostly as this is, the hook returns the necessary event handlers/props to attach to the element being wrapped.

For this case though, since our content is *almost always* block content (rather than e.g. spans), the wrapper seems a lot easier. Looking at Discord's implementation (you can use React dev tools in the browser one), it seems like they allow the wrapper behavior to be turned on or off, not quite sure why they need that complexity but 🤷